### PR TITLE
Use config colors to theme Wayland decorations

### LINKED ISSUE #2092: Take advantage of winit's new Wayland theming capability
In tomaka/winit#775, support for theming the title bar was added for Wayland for better UI integration.  Can Alacritty expose support for this through the config file (even if in some limited way)?  That would be awesome!

### PR DESCRIPTION
**Pull request by kchibisov**
Fixes #2092.

### CODE REVIEW

File: alacritty/src/window.rs [hunk 1]
```diff
@@ -361,6 +366,27 @@ impl Window {
...
+        let theme = AlacrittyWaylandTheme {
+            foreground,
+            background,
+            dim_foreground,
+            hovered_close_icon,
+            hovered_minimize_icon,
+            hovered_maximize_icon,
+        };
+        self.window().set_wayland_theme(theme);
+    }
```
Comment (nixpulvis) [inline alacritty/src/window.rs]: @kchibisov I thought you said this wouldn't need to be handled inside of Alacritty. It seems unfortunate to need to make all these choices for our users. What if my applications should all use blue to maximize?

Comment (kchibisov) [reply alacritty/src/window.rs]: You don't have control over it anyways, since it's a client side decorations. Before we had random white Qt'ish color scheme, now we have something that matches content. You can't just draw GTK decorations, etc. Providing direct config option isn't an option, so we should derive it our of users colorscheme.

Comment (nixpulvis) [reply alacritty/src/window.rs]: I feel like I may be misremembering a conversation... but I was under the impression we were against CSDs, in favor of SSDs. I can maybe understand wanting an escape hatch from poorly configured SSDs, but I really wish it wasn't inside out codebase.

Here's what I'm remembering: https://github.com/alacritty/alacritty/pull/2047#issuecomment-461925986

Comment (kchibisov) [reply alacritty/src/window.rs]: We support both CSD and SSD this option is for CSD theming, since we control them. We do support CSD and we're not against them. So on sway you'll use SSD on GNOME CSD.

Comment (nixpulvis) [reply alacritty/src/window.rs]: Sorry if this is obvious, but where do we "support both CSD and SSD" ATM?

Comment (nixpulvis) [reply alacritty/src/window.rs]: FWIW, my main issue here is with the plan for configuration of things like this. I don't really want to block enhancements that make Alacritty look better for users, since people will be happier to use Alacritty if it looks good.

Comment (kchibisov) [reply alacritty/src/window.rs]: >Sorry if this is obvious, but where do we "support both CSD and SSD" at the moment?

SCTK handles them for us.

File: CHANGELOG.md [hunk 2]
```diff
@@ -22,6 +22,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - The default value for `draw_bold_text_with_bright_colors` is now `false`
 - Mirror OSC query terminators instead of always using BEL
 - Increased Beam, Underline, and Hollow Block cursors' line widths
+- Wayland decorations are now based on config colorscheme
```
Comment (chrisduerr) [inline CHANGELOG.md]:
```suggestion
- Wayland client side decorations are now based on config colorscheme
```

File: alacritty/src/window.rs [hunk 3]
```diff
@@ -361,6 +366,27 @@ impl Window {
         self.window().wayland_display()
     }
 
+    #[cfg(not(any(target_os = "macos", target_os = "windows")))]
+    pub fn set_wayland_theme(&mut self, config: &Config) {
```
Comment (chrisduerr) [inline alacritty/src/window.rs]: Do we need to pass `config` here, when all we access is `config.colors`?

File: alacritty/src/window.rs [hunk 4]
```diff
@@ -361,6 +366,27 @@ impl Window {
...
+        let hovered_close_icon = config.colors.normal().red;
+        let hovered_maximize_icon = config.colors.normal().green;
+        let hovered_minimize_icon = config.colors.normal().yellow;
+        let foreground = config.colors.primary.foreground;
+        let background = config.colors.primary.background;
+        let dim_foreground =
+            config.colors.primary.dim_foreground.unwrap_or_else(|| foreground * 0.66);
```
Comment (chrisduerr) [inline alacritty/src/window.rs]: Not a fan of this, since we duplicate the dimming code here.

It might be better to pass `&config.colors` and `term.colors` so we can do it properly?

File: alacritty/src/window.rs [hunk 5]
```diff
@@ -386,6 +412,95 @@ impl Window {
...
+                self.hovered_close_icon.g,
+                self.hovered_close_icon.b,
+            ],
+            WaylandThemeButtonState::Idle => {
+                [0xff, self.foreground.r, self.foreground.g, self.foreground.b]
+            },
+            WaylandThemeButtonState::Disabled => {
+                [0xff, self.dim_foreground.r, self.dim_foreground.g, self.dim_foreground.b]
+            },
+        }
```
Comment (chrisduerr) [inline alacritty/src/window.rs]: I wonder if we could maybe write a little helper for that which would shorten things down?

Like `self.button_icon_color(status, self.hovered_close_icon)`.

Could then use that for all three buttons with just a single line in each of the trait implementations.

File: alacritty/src/window.rs (hunk 1 above)
Comment (nixpulvis) [reply alacritty/src/window.rs]: OK, point taken. I'd like to propose moving these out into their own `wayland.rs` file somewhere however.

Comment (chrisduerr) [reply alacritty/src/window.rs]: I'd just keep it simple with a `wayland_theme.rs` file for now and if we ever need more platform specific modules, we can separate it more.

File: alacritty/src/display.rs [hunk 6]
```diff
@@ -142,6 +142,13 @@ impl Display {
         // Spawn window
         let mut window = Window::new(event_loop, &config, size)?;
 
+        #[cfg(not(any(target_os = "macos", windows)))]
+        {
+            if event_loop.is_wayland() {
+                window.set_wayland_theme(&config);
+            }
+        }
```
Comment (nixpulvis) [inline alacritty/src/display.rs]: Is there a reason these aren't marked with a `#[cfg(feature = "wayland")]`?

Comment (kchibisov) [reply alacritty/src/display.rs]: well, there's no reason to opt out specific features code, since it doesn't affect compilation.

File: alacritty/src/window.rs (hunk 4 above)
Comment (kchibisov) [reply alacritty/src/window.rs]: You don't have access to terminal colors and I don't really like adding an option to get them, maybe we should keep this simple for now?

Comment (chrisduerr) [reply alacritty/src/window.rs]: I'm fine with keeping it simple, I'm not fine with keeping in bug prone.

If we wanted to keep it simple, we'd have to remove the dimming.

Why do we not have access to the terminal colors, because this is executed right at startup?

Comment (kchibisov) [reply alacritty/src/window.rs]: I mean they are `private` right now for us. We do have access for them, but it's about adding `pub` or methods to get them. This is not about error prone or not, let's say I've just dimmed color by some value if it wasn't available. I could remove use of dim and just use `0.64 * foreground`.

I feel like we can use dark black though.

Comment (chrisduerr) [reply alacritty/src/window.rs]: > This is not about error prone or not, let's say I've just dimmed color by some value if it wasn't available. I could remove use of dim and just use 0.64 * foreground.

This absolutely is about introducing bugs. Because if I change the dim colors in our color code, but forget to update the wayland theme, that's a bug.

> I feel like we can use dark black though.

No. That doesn't work for light themes.

Comment (kchibisov) [reply alacritty/src/window.rs]: Oh yeah, we can use just `white`.

**Comment by kchibisov** [comment]
There was some interest in playing with colors from @chrisduerr , so leaving this `dim` issue up to him. The only option I see is to either add some simple algorithm to get nice dim or to use config dim and add a method to `Term` to get it.

**Comment by chrisduerr** [comment]
@kchibisov Just fyi, I have no plan to play with this in the near future.

I'm open to suggestions from others and I do not particularly care much about the way the theme actually looks as long as reasonable colors from the user's colorscheme is picked. I just cannot approve an approach that uses bug-prone copy/pasting rather than proper clean code reuse.

**Comment by kchibisov** [comment]
Updated to reuse dim factor from terminal.

File: alacritty_terminal/src/term/mod.rs [hunk 7]
```diff
@@ -1260,6 +1260,11 @@ impl<T> Term<T> {
         self.colors[NamedColor::Background]
     }
 
+    #[inline]
+    pub fn dim_foreground_color(&self) -> Rgb {
+        self.colors[NamedColor::DimForeground]
+    }
```
Comment (chrisduerr) [inline alacritty_terminal/src/term/mod.rs]: This is never used?

File: alacritty/src/window.rs [hunk 8]
```diff
@@ -361,6 +368,26 @@ impl Window {
...
     }
 
+    #[cfg(not(any(target_os = "macos", target_os = "windows")))]
+    pub fn set_wayland_theme(&mut self, colors: &TerminalColors) {
+        let hovered_close_icon = colors.normal().red;
+        let hovered_maximize_icon = colors.normal().green;
+        let hovered_minimize_icon = colors.normal().yellow;
+        let foreground = colors.primary.foreground;
+        let background = colors.primary.background;
+        let dim_foreground = colors.primary.dim_foreground.unwrap_or(foreground * DIM_FACTOR);
```
Comment (chrisduerr) [inline alacritty/src/window.rs]: Didn't you say that dim foreground isn't a good choice, since it doesn't work with light themes?

File: alacritty/src/window.rs [hunk 9]
```diff
@@ -33,14 +33,21 @@ use image::ImageFormat;
 #[cfg(not(any(target_os = "macos", windows)))]
 use x11_dl::xlib::{Display as XDisplay, PropModeReplace, XErrorEvent, Xlib};
 
-use alacritty_terminal::config::{Decorations, StartupMode, WindowConfig};
+use alacritty_terminal::config::{
+    Colors as TerminalColors, Decorations, StartupMode, WindowConfig,
```
Comment (chrisduerr) [inline alacritty/src/window.rs]: Why did you rename this import, especially to `TerminalColors`? There is no conflict, right? And if anything, aren't these the config colors?

File: alacritty/src/window.rs [hunk 10]
```diff
@@ -361,6 +368,26 @@ impl Window {
...
+        let dim_foreground = colors.primary.dim_foreground.unwrap_or(foreground * DIM_FACTOR);
+
+        let theme = AlacrittyWaylandTheme {
+            foreground,
+            background,
+            dim_foreground,
+            hovered_close_icon,
+            hovered_minimize_icon,
+            hovered_maximize_icon,
+        };
```
Comment (chrisduerr) [inline alacritty/src/window.rs]: There's a lot of crap in this function body just taking things from `colors` and putting it inside of `AlacrittyWaylandTheme`. Could we just make a `new` function for the theme which takes a color, so that `window.rs` is less impacted?

File: alacritty_terminal/src/term/mod.rs (hunk 7 above)
Comment (kchibisov) [reply alacritty_terminal/src/term/mod.rs]: I thought I've removed it. Thx.

File: alacritty/src/window.rs (hunk 8 above)
Comment (kchibisov) [reply alacritty/src/window.rs]: Well, users can define their own dim foreground and has a control over the scheme, so I don't see any problem here.

File: alacritty/src/window.rs [hunk 11]
```diff
@@ -33,6 +33,8 @@ use image::ImageFormat;
 #[cfg(not(any(target_os = "macos", windows)))]
 use x11_dl::xlib::{Display as XDisplay, PropModeReplace, XErrorEvent, Xlib};
 
+#[cfg(not(any(target_os = "macos", windows)))]
+use alacritty_terminal::config::Colors;
 use alacritty_terminal::config::{Colors, Decorations, StartupMode, WindowConfig};
```
Comment (cole-h) [inline alacritty/src/window.rs]: Oops.

```suggestion
use alacritty_terminal::config::{Decorations, StartupMode, WindowConfig};
```

File: alacritty/src/window.rs [hunk 12]
```diff
@@ -361,6 +366,13 @@ impl Window {
         self.window().wayland_display()
     }
 
+    #[cfg(not(any(target_os = "macos", target_os = "windows")))]
+    pub fn set_wayland_theme(&mut self, colors: &Colors) {
+        let theme = AlacrittyWaylandTheme::new(colors);
+
+        self.window().set_wayland_theme(theme);
```
Comment (chrisduerr) [inline alacritty/src/window.rs]:
```suggestion
        self.window().set_wayland_theme(AlacrittyWaylandTheme::new(colors));
```

File: alacritty/src/window.rs [hunk 13]
```diff
@@ -41,6 +43,9 @@ use alacritty_terminal::term::{SizeInfo, Term};
 use crate::config::Config;
 use crate::gl;
 
+#[cfg(not(any(target_os = "macos", windows)))]
+use crate::wayland_theme::AlacrittyWaylandTheme;
```
Comment (chrisduerr) [inline alacritty/src/window.rs]:
```suggestion
#[cfg(not(any(target_os = "macos", windows)))]
use crate::wayland_theme::AlacrittyWaylandTheme;
```

File: alacritty/src/event.rs [hunk 14]
```diff
@@ -542,6 +544,13 @@ impl<N: Notify + OnResize> Processor<N> {
                             processor.ctx.display_update_pending.font = Some(font);
                         }
 
+                        #[cfg(not(any(target_os = "macos", windows)))]
+                        {
+                            if processor.ctx.event_loop.is_wayland() {
+                                processor.ctx.window.set_wayland_theme(&config.colors);
+                            }
+                        }
```
Comment (chrisduerr) [inline alacritty/src/event.rs]: This is getting kinda big, maybe we should consider moving the `ConfigReload` handling to a separate method?

File: alacritty/src/display.rs [hunk 15]
```diff
@@ -142,6 +142,13 @@ impl Display {
         // Spawn window
         let mut window = Window::new(event_loop, &config, size)?;
 
+        #[cfg(not(any(target_os = "macos", windows)))]
+        {
+            if event_loop.is_wayland() {
+                window.set_wayland_theme(&config.colors);
+            }
+        }
```
Comment (chrisduerr) [inline alacritty/src/display.rs]: Can't we put this in `window.rs`?

File: alacritty/src/event.rs [hunk 16]
```diff
@@ -675,6 +654,44 @@ impl<N: Notify + OnResize> Processor<N> {
...
+    pub fn reload_config<T>(
+        path: &PathBuf,
+        processor: &mut input::Processor<T, ActionContext<N, T>>,
+    ) where
+        T: EventListener,
+    {
+        processor.ctx.message_buffer.remove_target(LOG_TARGET_CONFIG);
+        processor.ctx.display_update_pending.message_buffer = Some(());
+
+        if let Ok(config) = config::reload_from(&path) {
```
Comment (chrisduerr) [inline alacritty/src/event.rs:666]:
```suggestion
        if let Ok(config) = config::reload_from(&path) {
        let config = match config::reload(&path) {
            Ok(config) => config,
            Err(_) => return, // Maybe log to debug?
        }
```

File: alacritty/src/wayland_theme.rs [hunk 17]
```diff
@@ -0,0 +1,83 @@
+use glutin::platform::unix::{ButtonState, Theme as WaylandTheme};
+
+use alacritty_terminal::config::Colors;
+use alacritty_terminal::term::color::Rgb;
+use alacritty_terminal::term::color::DIM_FACTOR;
```
Comment (chrisduerr) [inline alacritty/src/wayland_theme.rs:5]:
```suggestion
use alacritty_terminal::term::color::{Rgb, DIM_FACTOR};
```

File: alacritty/src/window.rs [hunk 18]
```diff
@@ -158,6 +162,10 @@ impl Window {
                 if let Some(parent_window_id) = config.window.embed {
                     x_embed_window(windowed_context.window(), parent_window_id);
                 }
+            } else {
+                let theme = AlacrittyWaylandTheme::new(&config.colors);
+
+                windowed_context.window().set_wayland_theme(theme);
```
Comment (chrisduerr) [inline alacritty/src/window.rs:168]:
```suggestion
                let theme = AlacrittyWaylandTheme::new(&config.colors);
                windowed_context.window().set_wayland_theme(theme);
```

This is mostly just a suggestion based on my preference. Doesn't really matter either way.

File: alacritty/src/event.rs (hunk 16 above)
Comment (kchibisov) [reply alacritty/src/event.rs:666]: This is already logged upper.

Comment (kchibisov) [reply alacritty/src/event.rs:666]: Also, clippy is against your change(asks for `if let`), especially you take into account that we log in `config::reload` already, since it's used from multiple paths.

**Comment by kchibisov** [comment]
>Before merging, would you mind sharing a picture (or more) of how these are supposed to look? Would probably be good for documentation anyways.

bitmap

Current CSD decorations looks like:
bitmap(1)

File: alacritty/src/event.rs (hunk 16 above)
Comment (chrisduerr) [reply alacritty/src/event.rs:666]: As discussed in IRC, this should still be done.

Comment (kchibisov) [reply alacritty/src/event.rs]: was done.
